crypto/src/external_commit.rs

```diff
@@ -0,0 +1,243 @@
+// Wire
+// Copyright (C) 2022 Wire Swiss GmbH
+
+// This program is free software: you can redistribute it and/or modify
+// it under the terms of the GNU General Public License as published by
+// the Free Software Foundation, either version 3 of the License, or
+// (at your option) any later version.
+
+// This program is distributed in the hope that it will be useful,
+// but WITHOUT ANY WARRANTY; without even the implied warranty of
+// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the
+// GNU General Public License for more details.
+
+// You should have received a copy of the GNU General Public License
+// along with this program. If not, see http://www.gnu.org/licenses/.
+
+use core_crypto_keystore::CryptoKeystoreMls;
+use openmls::{
+    group::{GroupId, MlsGroup},
+    prelude::{MlsMessageOut, VerifiablePublicGroupState},
+};
+use openmls_traits::OpenMlsCryptoProvider;
+
+use crate::{
+    prelude::{MlsConversation, MlsConversationConfiguration},
+    CryptoError, CryptoResult, MlsCentral, MlsError,
+};
+
+impl MlsCentral {
+    pub async fn join_by_external_commit(
+        &self,
+        group_state: VerifiablePublicGroupState,
+    ) -> CryptoResult<(GroupId, MlsMessageOut)> {
+        let credentials = self.mls_client.credentials();
+        let (mut group, message) = MlsGroup::join_by_external_commit(
+            &self.mls_backend,
+            None,
+            group_state,
+            &MlsConversationConfiguration::default().as_openmls_default_configuration(),
+            &[],
+            credentials,
+        )
+        .await
+        .map_err(MlsError::from)
+        .map_err(CryptoError::from)?;
+
+        let mut buf = vec![];
+        group.save(&mut buf)?;
```

Comment:
but this object is discarded in the end... the state is changed but after it serializes it. The method then just drops this object.

Reply:
Yes I know. I  never said it does not work (in fact it does), I said it's misleading.
This method is aimed at serializing a group for being persisted in a database. We want to serialize for crossing ffi boundaries: we should own this. We might decide that instead of json we want the group to be serialized in xml for whatever reason.
Also, tomorrow, this save method might decide (for its own reasons which might be perfectly valid) to mutate the group before serializing it.
The long-term clean solution is to handle ourselves the (de)serialization, it shouldn't be tied to openmls